Pr/add projects section to overview

HuntTheSun opened the pull request:
Hi,

It took me a bit, but I now have a first draft of the project spoiler feature. I anticipate that some modifications will be needed.

I tried to reuse and modularize existing code where possible, but I’m not entirely sure if I struck the right balance between refactoring and code duplication. I could probably modularize more, but I wanted to check in first to ensure the general direction is okay before making larger changes.

Some additions could be better integrated into the existing flow, but I wasn’t sure how to do so without major refactors. I’m happy to tackle those after we discuss what’s acceptable to refactor.

Could you confirm if the resulting UI is roughly what you envisioned?
[debug_dashboard.html](https://github.com/user-attachments/files/23415788/debug_dashboard.html)

### Commit categories chronologically:
#### Create data for runs already grouped by project in python 
Commit: feat(py/dashboard): add project grouped runs dict to dashboard data  
- Generates a dict of runs grouped by project and tag, available in JS.
- Avoids grouping at runtime in JS to save resources.
Specific Questions:
1. Is it alright if I add this extra pre-grouped data?
2. Should this dict be an extra key in the existing data instead of a separate dict? (or something entirely different)
3. Should this logic move to the database module instead?

#### Modularise existing functions
Commits: refactor(js)*
- Adapted existing logic to be reusable for this feature.

#### Add feature functions
Commits: feat(overview/projects)*
- Implemented "leaf" callee functions first, then moved up the call tree to their callers
- Each commit contains only already-defined functions.

Sorry for the open and closing of the PR, I accidentally posted it and panicked :)

HuntTheSun committed:
refactor(js): modularize "reset filter" and "set tagfilter selected project" as named functions

For also using the methods in the click listeners for the project run cards

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun committed:
refactor(js): convert anonymous compare_to_average function to named

Because I will need the function for the project runs cards also

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun committed:
refactor(js): add parameter to limit queryScope for setup_collapsables

**Why?**
When the projects section in overview gets recreated (for example when
toggling "Use Run Tags"), I dont want to query for and recreate all collapsables in
the whole document.

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun committed:
refactor(js): modularize overview run card generation to named function

**Why:**
I will need the function also for the run cards per project

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun committed:
refactor(js): modularize overview donut chart creation

**Why:**
I will also need that function for the donut charts in the projects
section of the overview

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun committed:
feat(overview/projects): add function to update card compares styling

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun committed:
feat(overview/projects): add html elements for projects section

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun committed:
feat(overview/projects): add project bars function

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun committed:
feat(overview/projects): add function for run card creation

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun committed:
feat(overview/projects): add function for run cards container creation

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun committed:
feat(overview/projects): add function for projects section creation

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

timdegroot1996 commented:
Alright nice work and you did it quick too! I don't have time to look at the code in depth at the moment so I will check those changes later and discuss refactoring in that comment. Here are my initial thoughts.

Resulting UI:

1. Good: I like the solution direction you chose by adding more layers of the overview statistics! This looks clean and it's clear to understand for users.
2. Improvement idea: I think it kind of duplicates functionality that we have the "Display All Runs" toggle still available in "Overview Statistics" because enabling this toggle creates exactly the same view as the added "Spoilers". Also the dropdown that it creates is basically the "choose your spoiler" when we in fact have them all below. Maybe we can completely remove this feature here as the spoilers kind of take over that functionality.
3. Question: The "Use Run Tags" toggle still has it's use and I think you handled how it updates the spoilers pretty nicely. I am debating if it would make sense to also add this toggle on every spoiler or if it is more logical to only keep it at the "Overview Statistics" part as it is now. Open to suggestions here, but keeping it as is wouldn't be bad in my opinion.

I saw these questions as well, so here are my answers.

1. Is it alright if I add this extra pre-grouped data?

I quickly looked at your code for this and I'm split on what would be the "nicer" solution. There are are a couple things that are to be taken into account:

- The data is now duplicated in the dashboard, so it will be bigger. Because the projects variable contains the same data as the run variable (to some extent)
- The dashboard will ALWAYS take extra time during the creation as it will always have to calculate this extra projects variable. (I recognize this is pretty minimal but it is a factor)
- If we move the logic to the "opening" of the overview page the logic would ALWAYS take the js calculating time (but I feel this shouldn't be too much either), and the bonus is that it ONLY does this if people actually go to the overview page. If people don't visit it there's no additional load.

With my current project architecture I've opted for "all logic in the dashboard" since this prevents data duplication. The dashboard is already quite big quite fast. I am all for performance optimizations but maybe this could be done by calculating the projects on load so it's only done once instead of repeatedly on tab opening to still gain performance?

2. Should this dict be an extra key in the existing data instead of a separate dict? (or something entirely different)

My answer above greatly influences this question. In line with the proposed solution there I suppose it could be a new variable but created on runtime, instead of added from python.

3. Should this logic move to the database module instead?

If we keep it the way it is now then I suppose that would be cleaner. But as I said above I might prefer not having duplicate data in the dashboard.

**Summary:** I'm curious what you think about mentioned aspects and if you agree with my conclusions. Code review will follow, but maybe it's more efficient to do this after we have answered all points above and implemented/refactored changes!

HuntTheSun commented:
> What the problem with the current (or slice(-1)) logic is that when the run you are slicing is not the last run. This can happen in the scenario where you are displaying all runs for a certain name or tag

Sorry, I still don't get why in this scenario the wrong duration should be sliced.
Arent the runs in the runList already correctly ordered from most recent to oldest, since the run cards always seem to be in the proper order?
In all project settings I tested from the example db, the most recent run was sliced away (which, as I understand, would be correct).
<details><summary>Debug Output</summary>
<pre><code class="language-js">
// log statements
                    console.log(runList);
                    const durationsForAvg = runList.map(x => x.duration).slice(1);
                    console.log(durationsForAvg);
                    const average = durationsForAvg.length ? avg(durationsForAvg) : duration;
                    console.log(average);
// output
Array(15) [ {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, … ]
0: Object { stats: (5) […], duration: "17.723", passed_run: 0 } //most recent run
1: Object { stats: (5) […], duration: "15.463", passed_run: 0 }
2: Object { stats: (5) […], duration: "23.442", passed_run: 0 }
3: Object { stats: (5) […], duration: "30.136", passed_run: 0 }
4: Object { stats: (5) […], duration: "14.655", passed_run: 0 }
5: Object { stats: (5) […], duration: "28.637", passed_run: 1 }
6: Object { stats: (5) […], duration: "17.016", passed_run: 1 }
7: Object { stats: (5) […], duration: "28.665", passed_run: 0 }
8: Object { stats: (5) […], duration: "13.148", passed_run: 0 }
9: Object { stats: (5) […], duration: "19.753", passed_run: 0 }
10: Object { stats: (5) […], duration: "14.333", passed_run: 0 }
11: Object { stats: (5) […], duration: "20.052", passed_run: 0 }
12: Object { stats: (5) […], duration: "26.243", passed_run: 0 }
13: Object { stats: (5) […], duration: "21.152", passed_run: 0 }
14: Object { stats: (5) […], duration: "14.429", passed_run: 0 }

//--------------------------------------------------------------------------------
Array(14) [ "15.463", "23.442", "30.136", "14.655", "28.637", "17.016", "28.665", "13.148", "19.753", "14.333", … ]
//--------------------------------------------------------------------------------
20.508857142857135 //average
</code></pre>
Check the math with python:
<pre><code class="language-py">
 sliced_durations = [
...   15.463,
...   23.442,
...   30.136,
...   14.655,
...   28.637,
...   17.016,
...   28.665,
...   13.148,
...   19.753,
...   14.333,
...   20.052,
...   26.243,
...   21.152,
...   14.429
... ]
>>> 
>>> sum(sliced_durations)
287.124
>>> sum(sliced_durations)/len(sliced_durations)
20.508857142857146  # correct
</code></pre>
</details>


Another thing I dont really get:
```js
sum(12, 10, 100, 10) - 12 === sum(10,100,10) // -> true , so why include 12 just to remove it?
// if you already know to subtract 12, it implies that you already know 12 is the correct element to remove,
// so why not remove it before sum(durations)
```

> Again, thanks a lot for your work. It's very motivating to see the engagement and progress 👍  

No, thank you for providing this awesome project, it is a joy to work with you.

> PS: I just learned the quoting part which is very neat hahaha  

Yeah it really comes in handy, I love me some markdown :)

HuntTheSun committed:
add function to group runs by project

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun committed:
remove "Show all Runs" switch

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun committed:
overview: add "Display by Name" switch

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun commented:
Hi, I hope you had a nice weekend.
> Yes that is exactly what I think it should be and your solution looks nice! Let's proceed with this.

Nice, I will add a commit for this.

> I saw some commits above, have you updated the stuff we discussed in my first comment? And is it ready for review again?

Not yet, it mostly works but the performance is not the best with a ton of results, I will try to make the projects section update
spoiler by spoiler, not the whole section. Its not as bad with the example db, but still noticeable.

Before I try to make it performant, is this roughly how the new Overview Buttons should work?
[debug_dashboard.html](https://github.com/user-attachments/files/23453271/debug_dashboard.html)

HuntTheSun committed:
overview: toggle project bar visibility

- create_project_overview now generates all project bars on first visit.
- toggling the switches now only hides/unhides project bars instead of
regeneration of whole projects section

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun committed:
overview: exclude self from run duration avg calc

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

timdegroot1996 commented:
> Hi, I hope you had a nice weekend.

I did have a nice weekend! A lot of personal stuff so unfortunately not as much time for robotdashboard as I hoped. But it was refreshing. I hope your weekend was nice too!

> Before I try to make it performant, is this roughly how the new Overview Buttons should work?

I tried it out and to be honest I didn't envision it like this but your solution matches your wording so I just understood differently. I envisioned it to be 1 button like it was and you toggle between views. I like your solution much better because it gives more insight and you can show more spoilers like this! I'm all for this solution 👍.

If you need help with the performance let me know, I might be able to help. I'm also quite curious what testdata you're using that is giving you performance issues.

Also don't forget to update this description that I'm hovering on the image below. I was checking if it was still matching the buttons and I feel like this might otherwise be forgotten:
<img width="791" height="693" alt="image" src="https://github.com/user-attachments/assets/2c70049e-a4a9-4a69-91db-a9fbc50c0e61" />

Edit: my god why is the image so insanely big hahahaha

HuntTheSun committed:
overview: fix project duration comparison update

Had to do index based duration element removal,
because run cards only store rounded duration

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun committed:
overview: update statistics info button text

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun committed:
overview: extract runs per row to global var

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun committed:
overview: extract default percentage to global var

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun committed:
overview: allow both "Display By" switches off

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun committed:
html: fix formatting

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

timdegroot1996 commented:
Hi again!

This morning I saw a lot of emails with commits so it seems you've been busy!
Thanks for letting me know that the PR is ready. I will try to find some time this afternoon or tonight to do a review of the code and the solution. I will also check it out and try to run it locally to see if there's anything we might have missed.

I'm not entirely sure how I will review it, it might either be comments in the actual pull request, or a comment here like I did before, but you will get notifications somehow when I'm ready.

HuntTheSun commented on robotframework_dashboard/templates/dashboard.html:
```diff
@@ -4124,80 +4335,74 @@ <h6 class="mb-0">Keyword Table</h6>
+            return;
+        }
+
+        function set_filter_show_current_project(projectName) {
```
Hi, 
if just ran a quick check, and it seems with the suggested simplification, clicking on the run card of a project based on run.name would not work.
It would show zero runs, as selectedRunSetting would not be updated, therefore the project name based filter would not apply and the filter would include no runs, because the tag-based filter doesn't work with non-tag project values.
```
        function set_filter_show_current_project(projectName) {
            // if (projectName.startsWith("project_")) {
                selectedTagSetting = projectName;
            // } else {
                // selectedRunSetting = projectName;
            // }
            return;
        }
```
[debug_dashboard.html](https://github.com/user-attachments/files/23477407/debug_dashboard.html)

timdegroot1996 commented on robotframework_dashboard/templates/dashboard.html:
```diff
@@ -4124,80 +4335,74 @@ <h6 class="mb-0">Keyword Table</h6>
-                    if (!foundTag) name = run.name;
-                } else {
-                    name = run.name;
+                    const tag = run.tags.split(",").find(t => t.startsWith("project_"));
```
This line seems to contain a bug when there are no tags at all. Which I found in some tests I just did locally. In those cases it returns "undefined" as the tag and you will see an "undefined" in the "Overview Statistics" if "Display by Tag" is enabled. I think something like this could be a clean solution:

```
const tag = run.tags.split(",").find(t => t.startsWith("project_"));
tag && projects.add(tag);
```

HuntTheSun replied:
Hi,

great catch, I didn't think of that at all, thank you.
That also got me thinking, what if instead of none, someone supplies two or more project_ tags?
This should handle 0-n project_ tags.
[commit](https://github.com/HuntTheSun/robotframework-dashboard/commit/40e2bbbd7e126fb11ff2b2364f154aa28dda64a4)
```diff
-                    const tag = run.tags.split(",").find(t => t.startsWith("project_"));
-                    projects.add(tag);
+                    const projectTags = run.tags.split(",").filter(t => t.startsWith("project_"));
+                    projectTags?.forEach(tag => projects.add(tag));
```
Are multiple project_ tags meant to be supported?
If not, I will write a commit with your suggested changes.
Here are some dashboards generated with the suggested edit.
[no_project_tags.html](https://github.com/user-attachments/files/23494506/no_project_tags.html)
[fix_undefined_and_multiple_projects.html](https://github.com/user-attachments/files/23494508/fix_undefined_and_multiple_projects.html) (each run has all 3 project_ tags)

timdegroot1996 replied:
What do you think about multiple project tags? Initially I would say only 1 is enough but I think your provided solution for multiple is actually fine too. It doesn't hurt I would say so maybe this solution is a nice improvement? I'll leave the final decision to you. You can close this comment at your discretion!

HuntTheSun replied:
Thank you for your thoughts.
I think the existing logic already has the ability to handle it, so it probably wouldn't hurt,
but I have no hard preference either way.
For the time being, I would keep multiple projects for now and if it causes unforeseen issues, rethink and adapt.

timdegroot1996 commented on robotframework_dashboard/templates/dashboard.html:
```diff
@@ -4300,48 +4477,53 @@ <h6 class="mb-0">Keyword Table</h6>
 
+        // ProjectNamePostfix is for differentiating overview statistic and project specific statistics
+        // since the project specific statistics are differentiated by adding project_ after projectName to the id
+        function generate_overview_card_html(projectName, stats, rounded_duration, status, runNumber, compares, passed_runs, log_path, log_name, svg, runIndex, projectNamePostfix = "") {
```
Like mentioned in the other comment at line ~4420 this argument "projectNamePostfix" can be removed because it can be solved differently.

timdegroot1996 commented on robotframework_dashboard/templates/dashboard.html:
```diff
@@ -1201,6 +1209,11 @@ <h5 class="modal-title" id="confirmModalLabel">Please Confirm</h5>
+        // populated by prepare_projects_grouped_data()
+        const projects_by_tag = {};
+        const projects_by_name = {};
+        let areGroupedProjectsPrepared = false; 
```
Just out of curiosity, why did you choose for let instead of var here? I'm not too experienced with the difference and sometimes I just do what chatgpt tells me 😄. I just know const can't be changed again, and var is global. "let" is still a mystery of its benefits to me. Maybe you could clarify!

HuntTheSun replied:
Hi,

I am in no way experienced in frontend/JS, 
but from what I gathered, it is generally recommended to use let if you can.
Let and var basically do the same, but var has some tricky definition scope behavior.
let is only defined within the scope that it is declared.
var on the other hand is defined within the whole function scope in which it is declared,
even if it is declared in a deeply nested block, it gets hoisted to the function scope.
Also, var can be used before it is declared.
I don't know if my explanation helped, here may be a better one:
https://sentry.io/answers/difference-between-let-and-var-in-javascript/

timdegroot1996 replied:
That does explain quite a bit. I don't really see benefits to using let over var in the global scope in these instances but on the other hand there's also no downside to using it. As it will simply update as it is in the "highest" scope. In loops or subfunctions I get why using let has more benefits now. Thanks for the explanation!

timdegroot1996 commented on robotframework_dashboard/templates/dashboard.html:
```diff
@@ -3976,6 +3967,226 @@ <h6 class="mb-0">Keyword Table</h6>
+            const logPath = use_logs ? transform_file_path(run.path).replaceAll('\\', '\\\\') : '';
+            const logName = use_logs ? 'log.html' : '';
+            const isDark = document.documentElement.classList.contains("dark-mode");
+            const svg = isDark ? clockDarkSVG : clockLightSVG;
```
I also found an interesting bug here. The bug is quite complicated to reproduce but I'll try to explain so you can try it yourself.

1. The first issue is that this code is only called once. So changing the theme to light doesn't actually update the clock icons in the new "spoilers"
2. This logic fires "too quickly" in some cases. What I mean by that is that the default them is "dark". So if you do the following:

- Set theme to light
- Refresh the page (ctrl + r)
- Open a spoiler -> the clock is white instead of the expected black svg
This happens because at first the page is in "dark mode" for a very short second, and then it gets updated with the local storage theme. But this happens AFTER you use this code.

Proposed solution is to add the clocks to the setup_theme function. That function is called after this function. If we add a class to all clocks svg's we could then update correclty based on the theme that is set. The reason the overview statistics clocks are not having this issue is because they are being generated after the localstorage is "taken into use" and when changing the theme they are regenerated by this last line of code:

```
function toggle_theme() {
            if (document.documentElement.classList.contains("dark-mode")) {
                set_local_storage_item("theme", "light");
            } else {
                set_local_storage_item("theme", "dark");
            }
            setup_theme()
            setup_dasbhoard_graphs()
        }
```

Solution suggestion:

1. Add this html in the card html creation function
```
                                        <div class="d-flex align-items-center ${compares}" id="${projectName}${projectNamePostfix}RunCardCompares${runIndex}">
                                            <div class="clock-icon">${svg}</div>
                                            <i id="${projectName}${projectNamePostfix}RunCardComparesValue${runIndex}" 
                                                value="${rounded_duration}" 
                                                class="ms-1">
                                                ${rounded_duration}s
                                            </i>
                                        </div>
```
2. Add these lines in the setup_theme function (light/dark respective). I added the move-up so you can find where it should be located in the code
```
                        ".move-up-section": moveUpLightSVG,
                        ".clock-icon": clockLightSVG,
                        
                        ".move-down-section": moveDownDarkSVG,
                        ".clock-icon": clockDarkSVG,
```
3. Unfortunately we can't completely take out the ${svg} because that would break the create_overview_statistics part as that is called AFTER the setup_theme logic in setup_data_and_graphs(). The setup_theme would therefore not populate the .clock-icons in that section as they are not yet created.

It's a bit of an ugly fix but if you have better suggestions I'm open to them.

HuntTheSun replied:
Hi,

thank you for the explanation, I was not familiar with the theme setup.
I cannot seem to reproduce refresh based clock icon bug, for me it either seems to work instantly or not at all,
I think I misinterpret what you mean.

But to make it work generally (excluding the "race bug" you described), I think your suggestion works.
I changed the suggested div to a span, because the div would cause a "linebreak".
```diff
--- a/robotframework_dashboard/templates/dashboard.html
+++ b/robotframework_dashboard/templates/dashboard.html
@@ -2533,0 +2534 @@ 
+                        ".clock-icon": clockLightSVG,
@@ -2595,0 +2597 @@ 
+                        ".clock-icon": clockDarkSVG,
@@ -4494 +4496,3 @@ 
-                                        ${svg}
+                                        <span class="clock-icon">
+                                            ${svg}
+                                        </span>
```
[commit](https://github.com/HuntTheSun/robotframework-dashboard/commit/def039a075d1a243ecd4f5f5742c6e7c90f62162)

Would that be an okay way?

I also just realized that the run donuts within the projects spoilers don't get updated on theme change,
I will look into it and fix that as well.

timdegroot1996 replied:
I tried out the fix-donut-graphs.html dashboard and it works perfectly. I'm just a little unsure of the reason you added the function update_donut_charts(scopeElement) function. Could you explain what it does? And why it is needed?

timdegroot1996 commented on robotframework_dashboard/templates/dashboard.html:
```diff
@@ -3976,6 +3967,226 @@ <h6 class="mb-0">Keyword Table</h6>
+                                        .reverse()
+                                        .filter((_, idx) => idx !== i)
+                const averageDuration = durationsForAvg.reduce((a,b) => a + parseFloat(b), 0) / durationsForAvg.length;
+                comparesElement.className = compare_to_average(duration, averageDuration, percentage);
```
This function would break my suggested solution as the class has gotten an additional "d-flex align-items-center ${compares}" to account for the clock changes. 

Simplest solution would be something like this:
```
 comparesElement.className = "d-flex align-items-center ${compare_to_average(duration, averageDuration, percentage)}";
```

HuntTheSun committed:
fix: prevent showing "undefined" if no project_

Also fix only first project_ project shown

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun committed:
refactor: remove projectNamePostfix

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun commented on robotframework_dashboard/templates/dashboard.html:
```diff
@@ -4209,37 +4414,30 @@ <h6 class="mb-0">Keyword Table</h6>
-                }
+                const runsToShow = [runList[0]];
 
                 runsToShow.forEach((r, runIndex) => {
```
Hi,

thank you for catching that, it seems I overlooked a lot of things to consider in the refactor.
I added a commit that hopefully resolves most of the refactors you mentioned pertaining to "projectNamePostfix".
[commit](https://github.com/HuntTheSun/robotframework-dashboard/commit/326c2eca27b961d20577750f5aa04f738be72440)

Sorry the commit contains so many changes, I had to adapt most html IDs.
Also, I adapted your suggested code above a little bit.
I hope I interpreted your comments correctly, please let me know what you think.
As always, no rush, I have enough work to do in the meantime :)

Thank you!

HuntTheSun committed:
remove dispensable svg theme check

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun committed:
fix overview clock svg update on theme change

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun committed:
prevent overwrite existing comparesValue classes

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun committed:
Revert "prevent overwrite existing comparesValue classes"

This reverts commit 757170191701dfb58899b43e2a2f4fc1091bfc3a.

HuntTheSun committed:
fix overview project run donuts not applying theme

Signed-off-by: Elias Haberl <ehaberl@tdt.de>

HuntTheSun committed:
fix typo setup_dashboard_graphs()

Signed-off-by: Elias Haberl <ehaberl@tdt.de>
